README.md
# denocryptor
An encryption and random data generation library for Deno, based on segcryptor for Node.js


CHANGELOG.md
## TODO

- implement `version`, `seed`, and `origin` in `makeID()`

---

## CHANGELOG

### 7 : renamed entry point
- renamed the project entrypoint, '/denocryptor.ts', to '/mod.ts'

### 6 : defaults fix
- fixed options to utilize a constructor for the class that loads a default value for each property if no value is provided

### 5 : sha512 implementation
- directly implemented the SHA-512 encryption algorithm in '/sha512.ts'

### 4 : optimized makeKey()
- made some minor optimizations to `makeKey()`

### 3 : options defaults
- updated options parameters to assume defaults if no value is provided

### 2 : updated tests
- updated '/src/tests.ts'

### 1 : initial commit
- initial commit


src/tests.ts
import {
  encryptPassword,
  hashValue,
  makeID,
  makeHash,
  makeKey,
  saltHash,
  toBoolean,
} from "./mod.ts";

/******************************************************************************/

function testEncryptPassword() {
  let hash =
    `d06ec4ed2e76e2c6c9e3dc3e1608735be0cbc0802ca5c2dac19bb492b8605708e3b8aff927d58a4c17fe9a7e6e3afcf3fc8c9e21791500ec0257b8fcca4c9d98`;
  let salt =
    `38537b935d7d0b98fe624ba268b6122b3b9b4d79fb0f714aecb1c55c7a5eb90eaba60f58f428c71117c036ccb7dfdfa80b002f99e96f3d0d36e64288a9b08fa9`;

  let encryptedPassword = encryptPassword(hash, salt, {
    count: 7, // how many times to run the hashing algorithm on the salted hash : high count = more secure
  });
  console.log(`testEncryptPassword : ${encryptedPassword}`);
}
testEncryptPassword();

/******************************************************************************/

function testHashValue() {
  let value = `you can hash pretty much any string or number`;

  let hashedValue = hashValue(value, {
    count: 5, // how many times you want to run the hashing algorithm : higher count = more secure
  });
  console.log(`testHashValue : ${hashedValue}`);
}
testHashValue();

/******************************************************************************/

function testMakeID() {
  let id = makeID({
    version: 1, // not yet implemented
    origin: ``, // not yet implemented
    seed: ``, // not yet implemented
    isSecure: true, // slower, but more secure random data generation
  });
  console.log(`testMakeID : ${id}`);
}
testMakeID();

/******************************************************************************/

function testMakeHash() {
  let randomData = makeHash({
    size: 128, // how many characters to generate
    toLowerCase: true, // produce results that consist of only lowercase hexadecimal characters (default)
    toUpperCase: false, // produce results that consit of only uppercase hexadecimal characters
    isSecure: false, // slower, but more secure random data generation
  });
  console.log(`testMakeHash : ${randomData}`);
}
testMakeHash();

/******************************************************************************/

function testMakeKey() {
  let randomKey = makeKey({
    size: 20,
    isComplex: false, // whether or not to generate both uppercase and lowercase alphanumeric characters
  });
  console.log(`testMakeKey : ${randomKey}`);
}
testMakeKey();

/******************************************************************************/

function testSaltHash() {
  let hash =
    `d06ec4ed2e76e2c6c9e3dc3e1608735be0cbc0802ca5c2dac19bb492b8605708e3b8aff927d58a4c17fe9a7e6e3afcf3fc8c9e21791500ec0257b8fcca4c9d98`;
  let salt =
    `38537b935d7d0b98fe624ba268b6122b3b9b4d79fb0f714aecb1c55c7a5eb90eaba60f58f428c71117c036ccb7dfdfa80b002f99e96f3d0d36e64288a9b08fa9`;
  let saltedHash = saltHash(hash, salt);
  console.log(`testSaltHash : ${saltedHash}`);
}
testSaltHash();

/******************************************************************************/

function testToBoolean() {
  let someNumber = 0;
  let numberToBool = toBoolean(someNumber);
  console.log(
    `testToBoolean : typeof numberToBool : ${typeof numberToBool} | value numberToBool : ${numberToBool}`,
  );

  let someWord = `true`;
  let wordToBool = toBoolean(someWord);
  console.log(
    `testToBoolean : typeof wordToBool : ${typeof wordToBool} | value wordToBool; : ${wordToBool}`,
  );
}
testToBoolean();


src/sha512.ts
/*
 * A JavaScript implementation of the Secure Hash Algorithm, SHA-512, as defined
 * in FIPS 180-2
 * Version 2.2 Copyright Anonymous Contributor, Paul Johnston 2000 - 2009.
 * Other contributors: Greg Holt, Andrew Kepert, Ydnar, Lostinet
 * Distributed under the BSD License
 * See http://pajhome.org.uk/crypt/md5 for details.
 */

// ! configurable variables
// ! you may need to tweak these to be compatible with server-side applications
// ! the defaults work in most cases
let hexcase = 0; /* hex output format. 0 - lowercase; 1 - uppercase        */
let b64pad = ``; /* base-64 pad character. "=" for strict RFC compliance   */

/******************************************************************************/

// create a 128-character, 512-bit hexidecimal hash string using SHA-512 encryption
function sha512(str: string) {
  return stringToHex(stringToSHA512(stringToUTF8(str)));
}
export { sha512 };

/******************************************************************************/

// create a base-64 string using SHA-512 encryption
function sha512B64(str: string) {
  return stringToBase64(stringToSHA512(stringToUTF8(str)));
}

/******************************************************************************/

// create a string hash using the SHA-512 encryption with any type of encoding
function sha512Encoded(str: string, enc: string) {
  return stringToEncoded(stringToSHA512(stringToUTF8(str)), enc);
}

/******************************************************************************/

// create a hexidecimal string of a HMAC-SHA-512
function hmacSHA512Hex(key: any, data: any) {
  return stringToHex(stringToHMACSHA512(stringToUTF8(key), stringToUTF8(data)));
}

/******************************************************************************/

// create a base-64 string of a HMAC-SHA-512
function hmacSHA512B64(key: any, data: any) {
  return stringToBase64(
    stringToHMACSHA512(stringToUTF8(key), stringToUTF8(data)),
  );
}

/******************************************************************************/

// create any encoding of a HMAC-SHA-512
function hmacSHA512Encoded(key: any, data: any, e: any) {
  return stringToEncoded(
    stringToHMACSHA512(stringToUTF8(key), stringToUTF8(data)),
    e,
  );
}

/******************************************************************************/

// calculate the SHA-512 hash of a raw string
function stringToSHA512(str: string) {
  return bigEndiansToString(
    bigEndiansToSHA512(stringToBigEndians(str), str.length * 8),
  );
}

/******************************************************************************/

// calculate the HMAC-SHA-512 of a key and some string of data
function stringToHMACSHA512(key: any, data: any) {
  let bkey = stringToBigEndians(key);
  if (bkey.length > 32) bkey = bigEndiansToSHA512(bkey, key.length * 8);

  const ipad = Array(32);
  const opad = Array(32);
  for (let i = 0; i < 32; i++) {
    ipad[i] = bkey[i] ^ 0x36363636;
    opad[i] = bkey[i] ^ 0x5c5c5c5c;
  }

  const hash = bigEndiansToSHA512(
    ipad.concat(stringToBigEndians(data)),
    1024 + data.length * 8,
  );
  return bigEndiansToString(bigEndiansToSHA512(opad.concat(hash), 1024 + 512));
}

/******************************************************************************/

// convert a raw string to a string of hexidecimal characters
function stringToHex(input: any) {
  if (hexcase === undefined) {
    hexcase = 0;
  }
  const hexTab = hexcase ? `0123456789ABCDEF` : `0123456789abcdef`;
  let output = ``;
  let x;
  const inputLength = input.length;
  for (let i = 0; i < inputLength; i++) {
    x = input.charCodeAt(i);
    output += hexTab.charAt((x >>> 4) & 0x0f) + hexTab.charAt(x & 0x0f);
  }
  return output;
}

/******************************************************************************/

// convert a raw string to a base-64 string
function stringToBase64(input: any) {
  if (b64pad === undefined) {
    b64pad = ``;
  }

  const tab =
    `ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/`;
  let output = ``;
  const len = input.length;
  for (let i = 0; i < len; i += 3) {
    const triplet = (input.charCodeAt(i) << 16) |
      (i + 1 < len ? input.charCodeAt(i + 1) << 8 : 0) |
      (i + 2 < len ? input.charCodeAt(i + 2) : 0);
    for (let j = 0; j < 4; j++) {
      if (i * 8 + j * 6 > input.length * 8) output += b64pad;
      else output += tab.charAt((triplet >>> (6 * (3 - j))) & 0x3f);
    }
  }
  return output;
}

/******************************************************************************/

// convert a raw string to a specific type of string encoding
function stringToEncoded(input: any, encoding: any) {
  const divisor = encoding.length;
  let i;
  let j;
  let q;
  let x;
  let quotient;

  /* Convert to an array of 16-bit big-endian values, forming the dividend */
  let dividend = Array(Math.ceil(input.length / 2));
  for (i = 0; i < dividend.length; i++) {
    dividend[i] = (input.charCodeAt(i * 2) << 8) | input.charCodeAt(i * 2 + 1);
  }

  // repeatedly perform a long division
  // the binary array forms the dividend,
  // the length of the encoding is the divisor
  // once computed, the quotient forms the dividend for the next step
  // all remainders are stored for later use
  const fullLength = Math.ceil(
    (input.length * 8) / (Math.log(encoding.length) / Math.log(2)),
  );
  const remainders = [fullLength];
  for (j = 0; j < fullLength; j++) {
    quotient = [];
    x = 0;
    for (i = 0; i < dividend.length; i++) {
      x = (x << 16) + dividend[i];
      q = Math.floor(x / divisor);
      x -= q * divisor;
      if (quotient.length > 0 || q > 0) {
        quotient[quotient.length] = q;
      }
    }
    remainders[j] = x;
    dividend = quotient;
  }

  /* Convert the remainders to the output string */
  let output = ``;
  const remaindersLength = remainders.length;
  for (i = remaindersLength - 1; i >= 0; i--) {
    output += encoding.charAt(remainders[i]);
  }

  return output;
}

// encode a string to utf-8
// this assumes the input string is valid utf-16
function stringToUTF8(input: any) {
  let output = ``;
  let i = -1;
  let x;
  let y;

  const inputLength = input.length;
  while (++i < inputLength) {
    /* Decode utf-16 surrogate pairs */
    x = input.charCodeAt(i);
    y = i + 1 < inputLength ? input.charCodeAt(i + 1) : 0;
    if (x >= 0xd800 && x <= 0xdbff && y >= 0xdc00 && y <= 0xdfff) {
      x = 0x10000 + ((x & 0x03ff) << 10) + (y & 0x03ff);
      i++;
    }

    /* Encode output as utf-8 */
    if (x <= 0x7f) {
      output += String.fromCharCode(x);
    } else if (x <= 0x7ff) {
      output += String.fromCharCode(
        0xc0 | ((x >>> 6) & 0x1f),
        0x80 | (x & 0x3f),
      );
    } else if (x <= 0xffff) {
      output += String.fromCharCode(
        0xe0 | ((x >>> 12) & 0x0f),
        0x80 | ((x >>> 6) & 0x3f),
        0x80 | (x & 0x3f),
      );
    } else if (x <= 0x1fffff) {
      output += String.fromCharCode(
        0xf0 | ((x >>> 18) & 0x07),
        0x80 | ((x >>> 12) & 0x3f),
        0x80 | ((x >>> 6) & 0x3f),
        0x80 | (x & 0x3f),
      );
    }
  }
  return output;
}

/******************************************************************************/

// encode a string to UTF-16, little endian
function stringToUTF16Little(input: any) {
  let output = ``;
  const inputLength = input.length;
  for (let i = 0; i < inputLength; i++) {
    output += String.fromCharCode(
      input.charCodeAt(i) & 0xff,
      (input.charCodeAt(i) >>> 8) & 0xff,
    );
  }
  return output;
}

/******************************************************************************/

// encode a string to UTF-16, big endian
function stringToUTF16Big(input: any) {
  let output = ``;
  const inputLength = input.length;
  for (let i = 0; i < inputLength; i++) {
    output += String.fromCharCode(
      (input.charCodeAt(i) >>> 8) & 0xff,
      input.charCodeAt(i) & 0xff,
    );
  }
  return output;
}

/******************************************************************************/

// convert a string to an array of big endian words
// characters greater than 255 have their high-byte ignored
function stringToBigEndians(input: any) {
  const inputLength = input.length;
  const output = Array(inputLength >> 2);
  const outputLength = output.length;
  for (let i = 0; i < outputLength; i++) {
    output[i] = 0;
  }
  for (let i = 0; i < inputLength * 8; i += 8) {
    output[i >> 5] |= (input.charCodeAt(i / 8) & 0xff) << (24 - (i % 32));
  }
  return output;
}

/******************************************************************************/

// convert an array of big endian words to a string
function bigEndiansToString(input: any) {
  let output = ``;
  const inputLength = input.length;
  for (let i = 0; i < inputLength * 32; i += 8) {
    output += String.fromCharCode((input[i >> 5] >>> (24 - (i % 32))) & 0xff);
  }
  return output;
}

/******************************************************************************/

// A constructor for 64-bit numbers
class int64 {
  h: number;
  l: number;
  constructor(h: number, l: number) {
    this.h = h;
    this.l = l;
  }

  // this.toString = int64toString;
}

// calculate the SHA-512 value of an array of big endian dwords, and a bit length
let sha512K: any;
function bigEndiansToSHA512(x: any, len: any) {
  if (sha512K == undefined) {
    // SHA512 constants
    sha512K = [
      new int64(0x428a2f98, -685199838),
      new int64(0x71374491, 0x23ef65cd),
      new int64(-1245643825, -330482897),
      new int64(-373957723, -2121671748),
      new int64(0x3956c25b, -213338824),
      new int64(0x59f111f1, -1241133031),
      new int64(-1841331548, -1357295717),
      new int64(-1424204075, -630357736),
      new int64(-670586216, -1560083902),
      new int64(0x12835b01, 0x45706fbe),
      new int64(0x243185be, 0x4ee4b28c),
      new int64(0x550c7dc3, -704662302),
      new int64(0x72be5d74, -226784913),
      new int64(-2132889090, 0x3b1696b1),
      new int64(-1680079193, 0x25c71235),
      new int64(-1046744716, -815192428),
      new int64(-459576895, -1628353838),
      new int64(-272742522, 0x384f25e3),
      new int64(0xfc19dc6, -1953704523),
      new int64(0x240ca1cc, 0x77ac9c65),
      new int64(0x2de92c6f, 0x592b0275),
      new int64(0x4a7484aa, 0x6ea6e483),
      new int64(0x5cb0a9dc, -1119749164),
      new int64(0x76f988da, -2096016459),
      new int64(-1740746414, -295247957),
      new int64(-1473132947, 0x2db43210),
      new int64(-1341970488, -1728372417),
      new int64(-1084653625, -1091629340),
      new int64(-958395405, 0x3da88fc2),
      new int64(-710438585, -1828018395),
      new int64(0x6ca6351, -536640913),
      new int64(0x14292967, 0xa0e6e70),
      new int64(0x27b70a85, 0x46d22ffc),
      new int64(0x2e1b2138, 0x5c26c926),
      new int64(0x4d2c6dfc, 0x5ac42aed),
      new int64(0x53380d13, -1651133473),
      new int64(0x650a7354, -1951439906),
      new int64(0x766a0abb, 0x3c77b2a8),
      new int64(-2117940946, 0x47edaee6),
      new int64(-1838011259, 0x1482353b),
      new int64(-1564481375, 0x4cf10364),
      new int64(-1474664885, -1136513023),
      new int64(-1035236496, -789014639),
      new int64(-949202525, 0x654be30),
      new int64(-778901479, -688958952),
      new int64(-694614492, 0x5565a910),
      new int64(-200395387, 0x5771202a),
      new int64(0x106aa070, 0x32bbd1b8),
      new int64(0x19a4c116, -1194143544),
      new int64(0x1e376c08, 0x5141ab53),
      new int64(0x2748774c, -544281703),
      new int64(0x34b0bcb5, -509917016),
      new int64(0x391c0cb3, -976659869),
      new int64(0x4ed8aa4a, -482243893),
      new int64(0x5b9cca4f, 0x7763e373),
      new int64(0x682e6ff3, -692930397),
      new int64(0x748f82ee, 0x5defb2fc),
      new int64(0x78a5636f, 0x43172f60),
      new int64(-2067236844, -1578062990),
      new int64(-1933114872, 0x1a6439ec),
      new int64(-1866530822, 0x23631e28),
      new int64(-1538233109, -561857047),
      new int64(-1090935817, -1295615723),
      new int64(-965641998, -479046869),
      new int64(-903397682, -366583396),
      new int64(-779700025, 0x21c0c207),
      new int64(-354779690, -840897762),
      new int64(-176337025, -294727304),
      new int64(0x6f067aa, 0x72176fba),
      new int64(0xa637dc5, -1563912026),
      new int64(0x113f9804, -1090974290),
      new int64(0x1b710b35, 0x131c471b),
      new int64(0x28db77f5, 0x23047d84),
      new int64(0x32caab7b, 0x40c72493),
      new int64(0x3c9ebe0a, 0x15c9bebc),
      new int64(0x431d67c4, -1676669620),
      new int64(0x4cc5d4be, -885112138),
      new int64(0x597f299c, -60457430),
      new int64(0x5fcb6fab, 0x3ad6faec),
      new int64(0x6c44198c, 0x4a475817),
    ];
  }

  // Initial hash values
  const H = [
    new int64(0x6a09e667, -205731576),
    new int64(-1150833019, -2067093701),
    new int64(0x3c6ef372, -23791573),
    new int64(-1521486534, 0x5f1d36f1),
    new int64(0x510e527f, -1377402159),
    new int64(-1694144372, 0x2b3e6c1f),
    new int64(0x1f83d9ab, -79577749),
    new int64(0x5be0cd19, 0x137e2179),
  ];

  const T1 = new int64(0, 0);
  const T2 = new int64(0, 0);
  const a = new int64(0, 0);
  const b = new int64(0, 0);
  const c = new int64(0, 0);
  const d = new int64(0, 0);
  const e = new int64(0, 0);
  const f = new int64(0, 0);
  const g = new int64(0, 0);
  const h = new int64(0, 0);

  // Temporary variables not specified by the document
  const s0 = new int64(0, 0);
  const s1 = new int64(0, 0);
  const Ch = new int64(0, 0);
  const Maj = new int64(0, 0);
  const r1 = new int64(0, 0);
  const r2 = new int64(0, 0);
  const r3 = new int64(0, 0);
  let j;
  let i;
  const W = new Array(80);
  for (i = 0; i < 80; i++) {
    W[i] = new int64(0, 0);
  }

  // append padding to the source string. The format is described in the FIPS.
  x[len >> 5] |= 0x80 << (24 - (len & 0x1f));
  x[(((len + 128) >> 10) << 5) + 31] = len;

  const xLength = x.length;
  for (
    i = 0;
    i < xLength;
    i += 32 // 32 dwords is the block size
  ) {
    int64copy(a, H[0]);
    int64copy(b, H[1]);
    int64copy(c, H[2]);
    int64copy(d, H[3]);
    int64copy(e, H[4]);
    int64copy(f, H[5]);
    int64copy(g, H[6]);
    int64copy(h, H[7]);

    for (j = 0; j < 16; j++) {
      W[j].h = x[i + 2 * j];
      W[j].l = x[i + 2 * j + 1];
    }

    for (j = 16; j < 80; j++) {
      // sigma1
      int64rrot(r1, W[j - 2], 19);
      int64revrrot(r2, W[j - 2], 29);
      int64shr(r3, W[j - 2], 6);
      s1.l = r1.l ^ r2.l ^ r3.l;
      s1.h = r1.h ^ r2.h ^ r3.h;

      // sigma0
      int64rrot(r1, W[j - 15], 1);
      int64rrot(r2, W[j - 15], 8);
      int64shr(r3, W[j - 15], 7);
      s0.l = r1.l ^ r2.l ^ r3.l;
      s0.h = r1.h ^ r2.h ^ r3.h;

      int64add4(W[j], s1, W[j - 7], s0, W[j - 16]);
    }

    for (j = 0; j < 80; j++) {
      // Ch
      Ch.l = (e.l & f.l) ^ (~e.l & g.l);
      Ch.h = (e.h & f.h) ^ (~e.h & g.h);

      // Sigma1
      int64rrot(r1, e, 14);
      int64rrot(r2, e, 18);
      int64revrrot(r3, e, 9);
      s1.l = r1.l ^ r2.l ^ r3.l;
      s1.h = r1.h ^ r2.h ^ r3.h;

      // Sigma0
      int64rrot(r1, a, 28);
      int64revrrot(r2, a, 2);
      int64revrrot(r3, a, 7);
      s0.l = r1.l ^ r2.l ^ r3.l;
      s0.h = r1.h ^ r2.h ^ r3.h;

      // Maj
      Maj.l = (a.l & b.l) ^ (a.l & c.l) ^ (b.l & c.l);
      Maj.h = (a.h & b.h) ^ (a.h & c.h) ^ (b.h & c.h);

      int64add5(T1, h, s1, Ch, sha512K[j], W[j]);
      int64add(T2, s0, Maj);

      int64copy(h, g);
      int64copy(g, f);
      int64copy(f, e);
      int64add(e, d, T1);
      int64copy(d, c);
      int64copy(c, b);
      int64copy(b, a);
      int64add(a, T1, T2);
    }
    int64add(H[0], H[0], a);
    int64add(H[1], H[1], b);
    int64add(H[2], H[2], c);
    int64add(H[3], H[3], d);
    int64add(H[4], H[4], e);
    int64add(H[5], H[5], f);
    int64add(H[6], H[6], g);
    int64add(H[7], H[7], h);
  }

  // represent the hash as an array of 32-bit dwords
  const hash = new Array(16);
  for (i = 0; i < 8; i++) {
    hash[2 * i] = H[i].h;
    hash[2 * i + 1] = H[i].l;
  }
  return hash;
}

// Copies src into dst, assuming both are 64-bit numbers
function int64copy(dst: any, src: any) {
  dst.h = src.h;
  dst.l = src.l;
}

// Right-rotates a 64-bit number by shift
// Won't handle cases of shift>=32
// The function revrrot() is for that
function int64rrot(dst: any, x: any, shift: any) {
  dst.l = (x.l >>> shift) | (x.h << (32 - shift));
  dst.h = (x.h >>> shift) | (x.l << (32 - shift));
}

// Reverses the dwords of the source and then rotates right by shift.
// This is equivalent to rotation by 32+shift
function int64revrrot(dst: any, x: any, shift: any) {
  dst.l = (x.h >>> shift) | (x.l << (32 - shift));
  dst.h = (x.l >>> shift) | (x.h << (32 - shift));
}

// Bitwise-shifts right a 64-bit number by shift
// Won't handle shift>=32, but it's never needed in SHA512
function int64shr(dst: any, x: any, shift: any) {
  dst.l = (x.l >>> shift) | (x.h << (32 - shift));
  dst.h = x.h >>> shift;
}

// Adds two 64-bit numbers
// Like the original implementation, does not rely on 32-bit operations
function int64add(dst: any, x: any, y: any) {
  const w0 = (x.l & 0xffff) + (y.l & 0xffff);
  const w1 = (x.l >>> 16) + (y.l >>> 16) + (w0 >>> 16);
  const w2 = (x.h & 0xffff) + (y.h & 0xffff) + (w1 >>> 16);
  const w3 = (x.h >>> 16) + (y.h >>> 16) + (w2 >>> 16);
  dst.l = (w0 & 0xffff) | (w1 << 16);
  dst.h = (w2 & 0xffff) | (w3 << 16);
}

// Same, except with 4 addends. Works faster than adding them one by one.
function int64add4(dst: any, a: any, b: any, c: any, d: any) {
  const w0 = (a.l & 0xffff) + (b.l & 0xffff) + (c.l & 0xffff) + (d.l & 0xffff);
  const w1 = (a.l >>> 16) +
    (b.l >>> 16) +
    (c.l >>> 16) +
    (d.l >>> 16) +
    (w0 >>> 16);
  const w2 = (a.h & 0xffff) +
    (b.h & 0xffff) +
    (c.h & 0xffff) +
    (d.h & 0xffff) +
    (w1 >>> 16);
  const w3 = (a.h >>> 16) +
    (b.h >>> 16) +
    (c.h >>> 16) +
    (d.h >>> 16) +
    (w2 >>> 16);
  dst.l = (w0 & 0xffff) | (w1 << 16);
  dst.h = (w2 & 0xffff) | (w3 << 16);
}

// Same, except with 5 addends
function int64add5(dst: any, a: any, b: any, c: any, d: any, e: any) {
  const w0 = (a.l & 0xffff) +
    (b.l & 0xffff) +
    (c.l & 0xffff) +
    (d.l & 0xffff) +
    (e.l & 0xffff);
  const w1 = (a.l >>> 16) +
    (b.l >>> 16) +
    (c.l >>> 16) +
    (d.l >>> 16) +
    (e.l >>> 16) +
    (w0 >>> 16);
  const w2 = (a.h & 0xffff) +
    (b.h & 0xffff) +
    (c.h & 0xffff) +
    (d.h & 0xffff) +
    (e.h & 0xffff) +
    (w1 >>> 16);
  const w3 = (a.h >>> 16) +
    (b.h >>> 16) +
    (c.h >>> 16) +
    (d.h >>> 16) +
    (e.h >>> 16) +
    (w2 >>> 16);
  dst.l = (w0 & 0xffff) | (w1 << 16);
  dst.h = (w2 & 0xffff) | (w3 << 16);
}


src/denocryptor.ts
import { sha512 } from "./sha512.ts";

/******************************************************************************/

class OptionsEncryptPassword {
  count: number; // default : 3 // number of times to apply the encryption algorithm to the result
  constructor(
    count?: number,
  ) {
    this.count = count || 3;
  }
}
export { OptionsEncryptPassword };

// applies a salt to a password hash and hashes it a predefined number of times equal to 'options.count'
function encryptPassword(
  hash: string, // the hash you're salting and then encrypting
  salt: string, // the salt hash that's going to be applied to the original hash, generally stored separately on the server for later uses
  options = new OptionsEncryptPassword(),
): string {
  let result: string = ``;

  // salt the hash, then run the SHA-512 encryption algorithm on the salted hash
  result = saltHash(hash, salt);
  result = hashValue(result, options);

  return result;
}
export { encryptPassword };

/******************************************************************************/

class OptionsHashValue {
  count: number; // default : 1 // how many times to apply the hashing algorithm to the 'value' string
  constructor(
    count?: number,
  ) {
    this.count = count || 1;
  }
}
export { OptionsHashValue };

// accepts a value, converts it to a string, and then applies the SHA-512 hash to it
function hashValue(
  value: any,
  options = new OptionsHashValue(),
): string {
  let result: string = (value as string);

  // run `options.count` number of iterations of the hashing algorithm
  for (let i = 0; i < options.count; i += 1) {
    result = (sha512(result) as string);
  }

  return result;
}
export { hashValue };

/******************************************************************************/

class OptionsMakeID {
  version: number; // default : 1 // TODO implement this
  origin: string | number; // default : Date.now() // TODO implement this
  seed: string | number; // default : Date.now() // TODO implement this
  isSecure: boolean; // default : false // whether or not to use the slower but more secure method of random data generation
  constructor(
    version?: number,
    origin?: string | number,
    seed?: string | number,
    isSecure?: boolean,
  ) {
    this.version = version || 1;
    this.origin = origin || (Date.now() as unknown as string);
    this.seed = seed || (Date.now() as unknown as string);
    this.isSecure = isSecure || false;
  }
}
export { OptionsMakeID };

// generate a random pseudo-UUID string
function makeID(
  options = new OptionsMakeID(),
): string {
  let result: string = ``;

  // TODO add different versions
  // TODO implement origin
  // TODO implement seed

  const characters = "0123456789abcdef";
  // generate the ID using the fast mode
  if (options.isSecure === false) {
    for (let i = 0; i < 36; i += 1) {
      let currentChar = `-`;
      if (i === 8 || i === 13 || i === 18 || i === 23) {
        currentChar = `-`;
      } else {
        let rando = Math.ceil(Math.random() * 256);
        rando %= 16;
        currentChar = characters[rando];
      }
      result += currentChar;
    }
  } else { // generate the ID using the slow mode
    let finished = false;
    let byteArray = new Array(36);
    do {
      const currentByte = Math.floor(Math.random() * 36);
      if (byteArray[currentByte] === undefined) {
        if (
          currentByte === 8 || currentByte === 13 || currentByte === 18 ||
          currentByte === 23
        ) {
          byteArray[currentByte] = `-`;
        } else {
          // generate a random byte to fill in the empty slot
          let rando = Math.ceil(Math.random() * 256);
          rando %= 16;
          byteArray[currentByte] = characters[rando];
        }
      }
      // check to see if all the bytes are filled and we're finished
      for (let i = 0; i < byteArray.length; i += 1) {
        if (byteArray[i] === undefined) {
          finished = false;
          break;
        } else {
          finished = true;
        }
      }
    } while (finished === false);
    // build the string using each byte in the array
    for (let i = 0; i < byteArray.length; i += 1) {
      result += byteArray[i];
    }
  }

  return result;
}
export { makeID };

/******************************************************************************/

class OptionsMakeHash {
  size: number; // default : 128 // how many random hexadecimal characters to generate
  toLowerCase: boolean; // default : true // whether the result should be all lowercase letters and numbers
  toUpperCase: boolean; // default : false // whether the result should be all uppercase letters and numbers
  isSecure: boolean; // default : false // whether to generate the random hash using a slower, but more secure method
  constructor(
    size?: number,
    toLowerCase?: boolean,
    toUpperCase?: boolean,
    isSecure?: false,
  ) {
    this.size = size || 128;
    this.toLowerCase = toLowerCase || true;
    this.toUpperCase = toUpperCase || false;
    this.isSecure = isSecure || false;
  }
}
export { OptionsMakeHash };

// generates a random hexidecimal hash, where `size` is the number of random hexadecimal characters to generate
function makeHash(
  options = new OptionsMakeHash(),
): string {
  let result: string = ``;

  //
  const characters = `0123456789abcdef`;
  if (options.isSecure === false) {
    // if using fast/less secure random generation, do it the easy way
    for (let i = 0; i < options.size; i += 1) {
      let rando = Math.floor(Math.random() * 16);
      const currentChar = characters[rando];
      result += currentChar;
    }
  } else {
    // generate the hash using the slower but slightly more random way
    let finished = false;
    let byteArray = new Array(options.size);
    do {
      const currentByte = Math.floor(Math.random() * options.size);
      if (byteArray[currentByte] === undefined) {
        let rando = Math.floor(Math.random() * 16);
        byteArray[currentByte] = characters[rando];
      }
      for (let i = 0; i < byteArray.length; i += 1) {
        if (byteArray[i] === undefined) {
          finished = false;
          break;
        } else {
          finished = true;
        }
      }
    } while (finished === false);
    // build the result string using the byte array data
    for (let i = 0; i < byteArray.length; i += 1) {
      result += byteArray[i];
    }
  }

  if (options.toLowerCase === true) {
    result = result.toLowerCase();
  } else if (options.toUpperCase === true) {
    result = result.toUpperCase();
  }
  return result;
}
export { makeHash };

/******************************************************************************/

class OptionsMakeKey {
  size: number; // default : 20 // how many alphanumeric characters to generate
  isComplex: boolean; // default : false // whether to use both uppercase and lowercase alphanumeric characters
  constructor(
    size?: number,
    isComplex?: boolean,
  ) {
    this.size = size || 20;
    this.isComplex = isComplex || false;
  }
}
export { OptionsMakeKey };

// generate a random alphanumeric string
function makeKey(
  options = new OptionsMakeKey(),
): string {
  let result: string = ``;

  try {
    if (options === undefined || typeof options !== `object`) {
      options = {
        size: 20,
        isComplex: false,
      };
    } else {
      if (options.size === undefined) {
        options.size = 20;
      } else {
        options.size = Number(options.size);
      }
      if (options.isComplex === undefined) {
        options.isComplex = false;
      } else {
        options.isComplex = toBoolean(options.isComplex);
      }
    }
    if (options.size % 1 !== 0 || options.size < 1) {
      // error, size must be an integer greater than or equal to 1
      throw new Error(
        `The size options for the key generation function was not an integer (whole number) greater than or equal to 1.`,
      );
    }
  } catch (ex) {
    throw new Error(
      `An exception error occurred while attempting to parse the options for the key generation function : ${ex.message}`,
    );
  }

  // calculate various values for use in determining a random value
  let upperRange = 3844;
  if (options.isComplex === true) {
    upperRange = 62;
  } else {
    upperRange = 36;
  }

  // generate a random character and then append it to the `result` string
  const characters =
    `0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ`;
  for (let i = 0; i < options.size; i += 1) {
    let rando = Math.floor(Math.random() * upperRange);
    result += characters[rando];
  }

  return result;
}
export { makeKey };

/******************************************************************************/

// combines a two hashes, then runs them through the SHA-512 hashing algorithm
function saltHash(
  hash: string,
  salt: string,
): string {
  let result = ``;

  // determine the length of the longest string
  let longer = 0;
  if (hash.length > salt.length) {
    longer = hash.length;
  } else {
    longer = salt.length;
  }
  // combine the two strings, attempting to interleave them
  for (let i = 0; i < longer; i += 1) {
    if (i < hash.length) {
      result += hash[i];
    }
    if (i < salt.length) {
      result += salt[i];
    }
  }

  return result;
}
export { saltHash };

/******************************************************************************/

// converts a string or numeric value to a boolean true or false
function toBoolean(
  value: number | string | boolean,
): boolean {
  if (typeof value === `number`) {
    if (value === 1) {
      return true;
    }
    return false;
  }
  if (typeof value === `string`) {
    value = value.toLowerCase();
    if (value === `true` || value === `yes` || value === `y` || value === `1`) {
      return true;
    }
    return false;
  }
  if (typeof value === `boolean`) {
    return value;
  }
  throw new Error(
    `Error attempting to parse ${JSON.stringify(value)} as a boolean value.`,
  );
}
export { toBoolean };
